twill-core/src/main/java/org/apache/twill/internal/kafka/client/SimpleKafkaConsumer.java

```diff
@@ -273,17 +274,19 @@ private MessageCallback wrapCallback(final MessageCallback callback,
       final AtomicBoolean stopped = new AtomicBoolean();
       return new MessageCallback() {
         @Override
-        public void onReceived(final Iterator<FetchedMessage> messages) {
+        public long onReceived(final Iterator<FetchedMessage> messages) {
           if (stopped.get()) {
-            return;
+            return Long.MIN_VALUE;
           }
-          Futures.getUnchecked(executor.submit(new Runnable() {
+          return Futures.getUnchecked(executor.submit(new Callable<Long>() {
+            long nextOffset = Long.MIN_VALUE;
             @Override
-            public void run() {
+            public Long call() {
               if (stopped.get()) {
-                return;
+                return nextOffset;
               }
-              callback.onReceived(messages);
+              nextOffset = callback.onReceived(messages);
```

Comment:
I was assuming that this `onReceived` can be called multiple times. Therefore, at line 286 when stopped is set to true, messages processed in previous calls of `onReceived` should be skipped.